pkg/api/steve/projectresources/handlers.go

```diff
@@ -0,0 +1,853 @@
+	if tlsSecret == nil { // in dev mode, there is no CA
+		apiService.Spec.InsecureSkipTLSVerify = true
+	} else {
+		apiService.Spec.CABundle = tlsSecret.Data[corev1.TLSCertKey]
+	}
+	// special non-project namespace to "contain" namespaces not in a project
+	namespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: unscopedNamespace,
+			Annotations: map[string]string{
+				NamespaceAnnotation: "true",
+			},
+		},
+	}
+	// cluster role for listing resources in this APIGroup
+	role := rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: RoleName,
+			Annotations: map[string]string{
+				AuthzAnnotation: "unscoped",
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:     []string{"list"},
+				APIGroups: []string{Group},
+				Resources: []string{"*"},
+			},
+		},
+	}
+	// default rolebinding for the special unscoped namespace, the prtb controller will create similar rolebindings for project namespaces for specific users
+	roleBinding := rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      RoleName,
+			Namespace: unscopedNamespace,
+			Annotations: map[string]string{
+				AuthzAnnotation: "unscoped",
+			},
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind: "Group",
+				Name: "system:authenticated", // all users can access this endpoint initially, the handler will do a SAR check before returning resources
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     RoleName,
+		},
+	}
+	err = a.apply.ApplyObjects(&apiService, &namespace, &role, &roleBinding)
+	return service, err
+}
+
+func getDynamicClient(config *wrangler.Context) (dynamic.Interface, error) {
+	restConfig, err := config.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return dynamic.NewForConfig(restConfig)
+}
+
+// authenticateAPIServer is a middleware that verifies the client's certificates and checks that the client's CN is in the API server's allow list.
+// This fulfills the two conditions outlines in https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#kubernetes-apiserver-client-authentication:
+// > 1. The connection must be made using a client certificate that is signed by the CA whose certificate is in --requestheader-client-ca-file.
+// > 2. The connection must be made using a client certificate whose CN is one of those listed in --requestheader-allowed-names.
+// The http.Server must have a TLS config with ClientAuth: tls.RequestClientCert set in order for the http.Request to have the client cert populated.
+// Normally mTLS would be handled at the network session layer automatically during the client connection when tls.RequireAndVerifyClientCert
+// is set on the Server's tls.Config, but since this server performs multiple functions and not all of them require mTLS, we handle it in the
+// application layer in this middleware.
+func (a *authHandler) authenticateAPIServer(next http.Handler) http.Handler {
```

Comment:
I think you're technically right, I don't see anything in kubernetes that sets a default for these values. kubeadm, kind, k3s and rke set one, I think in practice most providers should set it. I could log a warning if it's not set and default to open?

I noticed in the docs "Note: You can set this option to blank as --requestheader-allowed-names="". This will indicate to an extension apiserver that any CN is acceptable." which this doesn't currently handle, will fix that.

Reply:
> I could log a warning if it's not set and default to open?

I don't think we can default to open. Near as I can tell, we get the information on which user is making the request from the request from k8s. If we can't tell that the requesting user is k8s, any user could hit the service directly and just assert that they are an admin user (by including that users name in the headers), which would allow them to do things like list core secrets and effect cluster takeover. LMK if you are seeing something that could prevent that. 

I think if we want to use this in this fashion, we need to be sure that every cluster sets these args and be willing to error out if we can't get the configmap as desired.